### PR TITLE
Compiler

### PR DESCRIPTION
**Pull request by AndreiShilkin**
- 'Create' page was developed. User can create new task on this page;
- CompilerManager was created to isolate implementation from CompilerController;
- Mechanism for checking results was upgraded. Now system uses sets of test cases (combination input and output data) for checking;
- 'Solve' page was upgraded: CodeMirror code editor was added on 'Solve' page, output of results was improved;
- wwwroot/lib was added to .gitignore. Content of it will be restored at the first build.

### CODE REVIEW

File: src/Axe/Controllers/CompilerController.cs [hunk 1]
```diff
@@ -46,80 +51,64 @@ public IActionResult Solve(int id)
...
-                        EmitResult result = compilation.Emit(stream);
-                        if (!result.Success)
-                        {
-                            IEnumerable<Diagnostic> failures = result.Diagnostics.Where(diagnostic =>
-                                diagnostic.IsWarningAsError ||
-                                diagnostic.Severity == DiagnosticSeverity.Error);
-
-                            foreach (Diagnostic diagnostic in failures)
+                        case CodeBlockResult.SUCCESS:
+                            break;
```
Comment (AleksandrSharykin) [inline src/Axe/Controllers/CompilerController.cs]: empty `case: break;` can be simply omitted; it leaves only one `case` block, so  `switch` can be replaced with `if`

File: src/Axe/Controllers/CompilerController.cs [hunk 2]
```diff
@@ -46,80 +51,64 @@ public IActionResult Solve(int id)
...
-                    };
-
-                    CSharpCompilation compilation = CSharpCompilation.Create(
-                        assemblyName: assemblyName,
-                        syntaxTrees: new[] { syntaxTree },
-                        references: references,
-                        options: new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
-
-                    using (var stream = new MemoryStream())
+                    model.SourceCode = CSharpSyntaxTree.ParseText(model.SourceCode).GetRoot().NormalizeWhitespace().ToFullString();
```
Comment (AleksandrSharykin) [inline src/Axe/Controllers/CompilerController.cs]: this operation should not be in controller - move it to manager

File: src/Axe/Controllers/CompilerController.cs [hunk 3]
```diff
@@ -46,80 +51,64 @@ public IActionResult Solve(int id)
...
-
-                            foreach (Diagnostic diagnostic in failures)
+                        case CodeBlockResult.SUCCESS:
+                            break;
+                        case CodeBlockResult.FAILED:
+                        case CodeBlockResult.ERROR:
+                            for (int i = 0; i < result.Item2.Length; i++)
                             {
-                                ModelState.AddModelError(diagnostic.Id, diagnostic.GetMessage());
+                                ModelState.AddModelError(i.ToString(), result.Item2[i]);
```
Comment (AleksandrSharykin) [inline src/Axe/Controllers/CompilerController.cs]: pass ModelState to manager with other parameters (in Request<> wrapper) and perform all logic in manager

File: src/Axe/Managers/CompileManager.cs [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+        public Tuple<CodeBlockResult, string[]> Solve(CodeBlockVm model)
+        {
+            try
+            {
+                CodeBlock codeBlock = context.CodeBlock
+                    .Include(cb => cb.TestCases)
+                    .First(cb => cb.Id == model.Id);
+
+                string assemblyName = "AxeCompilerAssembly" + Path.GetRandomFileName();
+                // Random file name is necessary because .NETCore doesn't allow to unload assembly
```
Comment (AleksandrSharykin) [inline src/Axe/Managers/CompileManager.cs]: try load/unload assembly in a separate Application Domain

File: src/Axe/Managers/ICompileManager.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    public interface ICompileManager
+    {
+        /// <summary>
+        /// Returns a list of code block available for current user
+        /// </summary>
+        /// <param name="request"></param>
+        /// <returns></returns>
+        Task<List<CodeBlockVm>> GetCodeBlocks();
+
+        Task<CodeBlockVm> GetById(int id);
```
Comment (AleksandrSharykin) [inline src/Axe/Managers/ICompileManager.cs]: Needs `<summary>` for method and params

File: src/Axe/Models/CodeBlock.cs [hunk 6]
```diff
@@ -17,21 +17,56 @@ public class CodeBlock
...
         /// </summary>
+        [Required]
         public string Task { get; set; }
     }
+
+    /// <summary>
+    /// Enum represents type of output
+    /// </summary>
+    public enum OutputTypeEnum
+    {
```
Comment (AleksandrSharykin) [inline src/Axe/Models/CodeBlock.cs]: `enum` name doesn't need "Enum" suffix. OutputType sounds like a broad name, I suggest to use `SupportedType` name

File: src/Axe/Managers/CompileManager.cs (hunk 4 above)
Comment (AndreiShilkin) [reply src/Axe/Managers/CompileManager.cs]: .NET Core doesn't support Application Domain: https://github.com/dotnet/standard/blob/master/docs/faq.md#is-appdomain-part-of-net-standard

File: src/Axe/Models/CodeBlock.cs [hunk 7]
```diff
@@ -17,21 +17,56 @@ public class CodeBlock
...
+        [Display(Name = "int")]
+        INT,
+        [Display(Name = "double")]
+        DOUBLE,
+        [Display(Name = "string")]
+        STRING,
+        [Display(Name = "int[]")]
+        INT_ARRAY,
+        [Display(Name = "double[]")]
+        DOUBLE_ARRAY,
```
Comment (AleksandrSharykin) [inline src/Axe/Models/CodeBlock.cs]: Don't do CAPS or `_` for names. Just `Int`, `String`, `DoubleArray`, `IntArray`

File: src/Axe/Models/CodeBlock.cs [hunk 8]
```diff
@@ -17,21 +17,56 @@ public class CodeBlock
...
+        [Display(Name = "string[]")]
+        STRING_ARRAY
+    }
+
+    /// <summary>
+    /// Enum represents result of execution code block
+    /// </summary>
+    public enum CodeBlockResult
+    {
+        SUCCESS,
```
Comment (AleksandrSharykin) [inline src/Axe/Models/CodeBlock.cs]: Don't do CAPS

File: src/Axe/Views/Compiler/Create.cshtml [hunk 9]
```diff
@@ -0,0 +1,136 @@
+@using Axe.Models
+@model Axe.ViewModels.CompilerVm.CodeBlockTaskVm
+
+@{
+    ViewData["Title"] = "Add new task";
+}
+
+@section Scripts
+{
+    <script type="text/javascript">
```
Comment (AleksandrSharykin) [inline src/Axe/Views/Compiler/Create.cshtml]: move JS scripts to separate files (.js) and reference them using `src`

File: src/Axe/Views/Compiler/Solve.cshtml [hunk 10]
```diff
@@ -1,32 +1,66 @@
-@model Axe.Models.CodeBlock
+@using Axe.Models;
+@model Axe.ViewModels.CompilerVm.CodeBlockVm
 @{
     ViewData["Title"] = "Solve task #" + Model.Id;
 }
 
+@section Scripts
+{
+    <script type="text/javascript">
```
Comment (AleksandrSharykin) [inline src/Axe/Views/Compiler/Solve.cshtml]: script to a separate file

**Review by AleksandrSharykin — CHANGES_REQUESTED** [review]
naming issues, separation of operation between controller and manager, script location - need fixing

File: src/Axe/Controllers/CompilerController.cs [hunk 11]
```diff
@@ -46,80 +51,59 @@ public IActionResult Solve(int id)
...
-                    CSharpCompilation compilation = CSharpCompilation.Create(
-                        assemblyName: assemblyName,
-                        syntaxTrees: new[] { syntaxTree },
-                        references: references,
-                        options: new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
-
-                    using (var stream = new MemoryStream())
+                    Tuple<CodeBlockResult, string[]> result = compileManager.Solve(model);
+                    model.Result = result.Item1;
+                    if (result.Item1 == CodeBlockResult.Error || result.Item1 == CodeBlockResult.Failed)
```
Comment (AleksandrSharykin) [inline src/Axe/Controllers/CompilerController.cs]: I'm still expecting when this post-processing is moved to manager before accepting pull-request

File: src/Axe/Managers/CompileManager.cs [hunk 12]
```diff
@@ -0,0 +1,437 @@
...
+                            break;
+                    }
+
+                    codeBlock.VerificationCode = codeBlock.VerificationCode
+                        .Replace("#INDEX#", i.ToString())
+                        .Replace("#OUTPUT#", codeBlock.TestCases[i].Output)
+                        .Replace("#INPUT#", codeBlock.TestCases[i].Input);
+                }
+
+                switch (codeBlock.OutputType)
```
Comment (AleksandrSharykin) [inline src/Axe/Managers/CompileManager.cs]: a lot of duplicated code in each `case`. there must be a shorter way to map SupportedType values to type name, e.g: read all SupportedType enum values, get `[Display]` attribute, store them in `Dictionary<SupportedType, string>` . then retrive a string based on `codeBlock.OutputType` and do `codeBlock.VerificationCode = codeBlock.VerificationCode.Replace(...)`  once

Comment (AndreiShilkin) [reply src/Axe/Managers/CompileManager.cs]: I fixed it. Now, I use Display attribute of SupportedType enum. See it in the commit: https://github.com/AleksandrSharykin/Axe/pull/19/commits/807c0047976b0e9035738b1a37e7094f6dc8b4b8

File: src/Axe/Controllers/CompilerController.cs (hunk 11 above)
Comment (AndreiShilkin) [reply src/Axe/Controllers/CompilerController.cs]: I already all implementation is moved to CompileManager. The Solve method returns Tuple where the first item is result, the second is string array of errors. Thus, you can use unit tests to test this method. In controllers I do the only action - to prepare model for view (add errors to ModelState if they are)

**Comment by AndreiShilkin** [comment]
Now everyone can solve, edit and delete tasks. Support of roles will be added in the future.

**Pull request merged**